Adapt Ada code and take advantage of new SPARK features

joffreyhuguet opened the pull request:
This patch makes the Ada code compatible with the changes done in GNAT and SPARK. It takes advantage of new features such as logical equality, both in containers and more generally in SPARK code, to remove lots of wrappers that were used to reduce context for provers to prove invariants on the service state.

manthonyaiello committed:
WIP Alire integration

Install working; need to make use of it in anod for printenv.

manthonyaiello committed:
Leverage the use of alr in `ensure_gnat`

This allows us to use gnat and gnatprove, after they've been installed
by Alire.

manthonyaiello committed:
Add XMLAda and fix auto-install.

Last commit broke auto-install of GNAT; this fixes it.

Also adds XMLAda from alr. It appears XMLAda was removed from the
alr-provided GNAT.

This patch is still broken, however, because of possible gprbuild
issues.

manthonyaiello committed:
Build XMLAda with alr.

joffreyhuguet committed:
Adapt Ada code and take advantage of new SPARK features

This patch makes the Ada code compatible with the changes done
in GNAT and SPARK. It takes advantage of new features such as
logical equality, both in containers and more generally in
SPARK code, to remove lots of wrappers that were used to
reduce context for provers to prove invariants on the
service state.

joffreyhuguet commented:
This is only a draft PR because the new features are not publicly available (in Community Edition or on alire).

joffreyhuguet committed:
Replace bounded formal containers by unbounded formal containers

[changelog]
* src/ada/src/services/
Replace bounded formal containers by unbounded formal containers in
Route aggregator and Assignment Tree Branch Bound. Fix proofs.

* src/ada/proof/sessions
Update session files.

* tests/proofs/
Update tests outputs.

manthonyaiello committed:
Fix spurious spaces in anod script

manthonyaiello committed:
Install xmlada via anod, rather than Alire.

This works around a problem with the Alire version not being
relocatable.

manthonyaiello committed:
Use infrastructure install for gnat

... rather than the GitHub action.

manthonyaiello committed:
Update Alire support for 13.2.1

Note: this commit is in an intermediate state, as we're waiting for
resolution of an Alire config-directory-handling error.

manthonyaiello committed:
Updates to enable successful build with 13.2.0

manthonyaiello committed:
Add commands to work around Alire bug.

manthonyaiello committed:
Add support for adding GNAT FSF to the path.

This will happen when we do `anod printenv uxas-ada --build-env`

manthonyaiello commented:
At time of writing, this PR resolves #72 and successfully builds the OpenUxAS Ada sources. Given the conflict, we'll have to do some rebasing / cleanup before this is ready to be merged.

But there's a bigger issue: the version of SPARK delivered in GNATprove FSF 13.2.0 is inappropriate for the state of the SPARK sources reflected in this PR - so we still don't have complete proofs. We expect that in moving to a newer of SPARK, these proof failures will be resolved automatically.

Leaving this PR as a draft until cleanup is complete (at least) and probably until we have a sufficient version of SPARK to resolve the proof failures.

manthonyaiello committed:
Use the nightly build of Alire for local config

The nightly build of Alire has the fix to use local configs for
toolchain selection. This removes the need for the workaround we put in
place on a prior commit.

Once there's a new release of Alire with this fix, we'll update again to
use that version, for stability.

joffreyhuguet committed:
Update code to get rid of linker errors using sparklib.gpr.
Update session files.
Update proof tests outputs.

manthonyaiello commented:
@VVCAS-Sean This is now in a state that's fit to be reviewed and merged. The failure on Ubuntu 20.04 is attributable to this issue, reported at Alire: https://github.com/alire-project/alire/issues/1526 - this will go away once we move away from Alire nightlies or if Alire changes the way it builds nightlies.
